dev_docs/key_concepts/persistable_state.mdx

```diff
@@ -0,0 +1,68 @@
+
+Example: Data plugin allows you to generate filters. Those filters can be persisted by applications in their saved 
+objects or in the URL. In order to allow apps to migrate the filters in case the structure changes in the future, the Data plugin implements `PersistableStateService` on `data.query.filterManager`.
+
+note: There is currently no obvious way for a plugin to know which state is safe to persist. The developer must manually look for a matching `PersistableStateService`, or ad-hoc provided migration utilities (as is the case with Rule Type Parameters).
+In the future, we hope to make it easier for producers of state to understand when they need to write a migration with changes, and also make it easier for consumers of such state, to understand whether it is safe to persist.
+
+## Exposing state that can be persisted but is not owned by plugin exposing it (registry)
+
+Any plugin that owns collection of items whose state/configuration can be persisted should implement `PersistableStateService`
+interface on their `setup` contract as well as make each item in the collection implement `PersistableStateDefinition` interface.
+
+Example: Embeddable plugin owns the registry of embeddable factories to which other plugins can register new embeddable factories. Dashboard plugin
+stores a bunch of embeddable panels input in its saved object and URL. Embeddable plugin setup contract implements `PersistableStateService`
+interface and each `EmbeddableFactory`  needs to implement `PersistableStateDefinition` interface.
+
+note: if your plugin doesn't expose the state and is thus the only one storing state but still owning the registry
+of items whose state it doesn't own then the items should still implement `PersistableStateDefinition` but registry
+plugin doesn't need to implement `PersistableStateService` interface.
+
+## Storing persistable state as part of saved object
+
+Any plugin that stores any persistable state as part of their saved object should make sure that its saved object migration 
+and reference extraction and injection methods correctly use the correct `PersistableStateService` implementation.
+
+If a saved object contains state from another plugin, the <DocLink id="kibCoreSavedObjectsPluginApi" section="iaf33fb81-e64a-11eb-8631-9d6e17a74ca4" text-"SavedObjectType.migrations"/> parameter should be a function, which returns a list of migrations. 
+ These migration functions will need to merge the migration functions returned from with any migrations provided by the plugin.
+
+For reference injection and extraction matching methods on PersistableStateService should be called and merged with plugins own array of references.
+
+## Available state operations
+
+### Storing persistable state as part of URL
+
+When storing persistable state as part of URL you must make sure your URL is versioned. When loading the state `migrateToLatest` method
```

Comment:
> When storing persistable state as part of URL you must make sure your URL is versioned

Do we have suggestions on how to version the URL? Is it simply adding some field that has a version value in it?

Reply:
nothing, its completely up to the app how it does that.